estate/models/estate_property.py

```diff
@@ -0,0 +1,103 @@
+from odoo.exceptions import UserError
+import datetime
+from dateutil.relativedelta import relativedelta
+from odoo.tools.float_utils import float_compare
+
+
+class EstateProperty(models.Model):
+    _name = "estate.property"
+    _description = "test description"
+    _order = "id desc"
+
+    name = fields.Char('Name', required=True)
+    description = fields.Text('Description')
+    postcode = fields.Char('Postcode')
+    date_availability = fields.Date('Date availability', copy=False, default=datetime.date.today() + relativedelta(months=+3))
+    expected_price = fields.Float('Expected price', required=True)
+    selling_price = fields.Float('Selling price', readonly=True, copy=False)
+    bedrooms = fields.Integer('Bedrooms', default=2)
+    living_area = fields.Integer('Living area (sqm)')
+    facades = fields.Integer('Facades')
+    garage = fields.Boolean('Garage')
+    garden = fields.Boolean('Garden')
+    garden_area = fields.Integer('Garden area (sqm)')
+    garden_orientation = fields.Selection(
+        string='Orientation',
+        selection=[('north', 'North'), ('east', 'East'), ('south', 'South'), ('west', 'West')],
+    )
+    active = fields.Boolean('Active', default=True)
+    state = fields.Selection(
+        string='state',
+        selection=[('new', 'New'), ('offerreceived', 'Offer Received'), ('offeraccepted', 'Offer Accepted'), ('sold', 'Sold'), ('cancelled', 'Cancelled')],
+        default='new',
+    )
+    total_area = fields.Integer('Total area (sqm)', compute='_compute_total_area')
+    best_offer = fields.Float('Best price', compute='_compute_best_offer')
+    property_type_id = fields.Many2one('estate.property.type', string="Property Type")
+    salesperson_id = fields.Many2one('res.users', string='Salesman', default=lambda self: self.env.user)
+    buyer_id = fields.Many2one('res.partner', string='Buyer', copy=False)
+    tag_ids = fields.Many2many('estate.property.tag', string="Property Tags")
+    offer_ids = fields.One2many('estate.property.offer', 'property_id', string="Offers")
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    @api.depends('offer_ids.price')
+    def _compute_best_offer(self):
+        for record in self:
+            record.best_offer = max(record.offer_ids.mapped('price')) if record.offer_ids else 0
+
+    @api.onchange('garden')
+    def _onchange_garden(self):
+        if self.garden:
+            self.garden_area = 10
+            self.garden_orientation = 'north'
+        else:
+            self.garden_area = 0
+            self.garden_orientation = None
+
+    def action_set_cancelled(self):
+        for record in self:
+            if record.state == 'sold':
+                raise UserError('A sold property cannot be cancelled')
+            else:
+                record.state = 'cancelled'
+        return True
```

Comment:
Here, your action is used on a form. That means that you are updating only one record. So you do not need to iterate `self` and you can add a `self.ensure_one()` to make sure that your self contains only one record, if it is not the case it triggered an error, and so it protects the db. You can apply that to all method called by a button in a form. 

```suggestion
        self.ensure_one()
        if self.state == 'sold':
                raise UserError('A sold property cannot be cancelled')
        record.state = 'cancelled'
        return True
```